### ISSUE TITLE
Accessing index out of bounds must give an index error

### ISSUE THREAD

**Issue by namannimmo10** [lcompilers/lpython#333]
The following code should not compile:
```python
def main0():
    s: str
    s = "abcdefg"
    print(s[0])
    print(s[12])

    arr: i32[4]
    arr[0] = 1
    arr[1] = 0
    arr[2] = -4
    arr[9] = 4

main0()
```

**Comment by certik** [comment]
Yes. There is compile time checking and runtime checking. We should design both to be language independent, so that the logic can be shared with LFortran.

I would suggest to first do a solid runtime checking. That would be done in the LLVM backend every time you reference an array. Just insert a function call `bounds_check` with the max dimension and the actual index, for each dimension.
That will give an error message at runtime for both cases above.

For compile time checking, we can either write an ASR pass, or simply check it in AST to ASR. But I would create a function in libasr, to which you pass the ASR node such as `arr[9]` and it will check the bounds (if available at compile time) and if it fails, give a compile time error message.

Also, both should be possible to turn off using a compile flag. And it should be on by default in Debug mode (which itself should be the default mode), and off by default in Release mode.

**Comment by czgdp1807** [comment]
To enable IndexError in LPython following should be done,

1. Uncomment and implement the following method,

https://github.com/lcompilers/lpython/blob/55a4efa7d983d2d40bd5b8e6375d302b5f93bfe4/src/libasr/codegen/llvm_array_utils.cpp#L413-L415

2. Uncomment the call to above method,

https://github.com/lcompilers/lpython/blob/55a4efa7d983d2d40bd5b8e6375d302b5f93bfe4/src/libasr/codegen/llvm_array_utils.cpp#L428-L430

3. Pass the value of `check_bounds` according to the compile flag as @certik pointed in the last line of his comment,

https://github.com/lcompilers/lpython/blob/55a4efa7d983d2d40bd5b8e6375d302b5f93bfe4/src/libasr/codegen/llvm_array_utils.cpp#L417-L419

**Comment by prathameshpatil2002** [comment]
Hi, I'd like to work on this issue, I've already setup the whole development environment an ran all the tests successfully. The only issue is that I'm unable to use the lpython executable to execute my  Pyhton scripts, due to lack of documentation I'm unable to resolve this small issue. Could any of you help me in using lpython, once that is done I'll try to resolve as many issues as I can.

**Comment by Smit-create** [comment]
I think this works now:
```console
% lpython a.py
a
String index out of Bounds
```
